client/messaging/connected-username.tsx

```diff
@@ -0,0 +1,55 @@
+import React, { useCallback, useRef, useState } from 'react'
+import styled from 'styled-components'
+import { ConnectedUserProfileOverlay } from '../profile/user-profile-overlay'
+import { useAppSelector } from '../redux-hooks'
+
+const Username = styled.span`
+  cursor: pointer;
+`
+
+/**
+ * A component which displays the username in messaging-related services (e.g. chat, lobbies,
+ * whispers) and any additional functionality that we might want to do with them, like opening a
+ * user profile when clicked.
+ *
+ * This component is connected to the store where it tries to find the user. All the services using
+ * it should ensure that the user is loaded in the store properly.
+ */
+export function ConnectedUsername(props: { userId: number }) {
+  const [overlayOpen, setOverlayOpen] = useState(false)
+  const usernameRef = useRef(null)
+
+  const onOpenOverlay = useCallback(() => {
+    setOverlayOpen(true)
+  }, [])
+  const onCloseOverlay = useCallback(() => {
+    setOverlayOpen(false)
+  }, [])
+
+  const user = useAppSelector(s => s.users.byId.get(props.userId))
+  if (!user) {
+    return <span>Unknown user</span>
+  }
+
+  return (
+    <>
+      <ConnectedUserProfileOverlay
```

Comment:
I worry a *lot* about introducing this here tbh. This is already kind of an issue with the user list, every single entry there introduces a Portal + managed div outside the react root, so this means every single message will do this as well (and messages are not virtualized so potentially this is a *lot* of elements to be managing). Would vastly prefer that this communicates with some other component (whether via context or redux actions or whatever) so we only maintain a single element for the whole containing view

Reply:
I tried to do this, and turns out it's not that easy at all. Unless I'm doing something wrong, which is also very likely.

So I tried doing this in the same way we're opening dialogs, through the store. My initial thinking was to save the popover props in the store and then have a connected popover somewhere in the app that would pull those props from the store and render the popover. However, it's pretty much impossible to make it work like this unless we watch for `anchorX/Y` changes in the component that calculates them, and we update the store each time they change.

The other option is calculating the anchor position inside the connected popover itself, but then we'd need to save in store enough information to make it possible to calculate (along with potential offsets), which I'm not a fan of at all.

So yeah, I'm kinda stuck here and not sure what's the best way to deal with it.

Another option might be trying to see if it's possible to change the Popover/Portal code to not create `div` elements _until_ the Popover is actually opened.

Reply:
I think Context is actually a better/easier API to do this with since you can pass non-serializable objects in, and have some easier JS API on top of it. If you use keys "properly" (not sure how easy this is to do with our current popover code, but ideally it would be simple :) ) then you could get proper show/hide animation even when people click on different users in succession.

> Another option might be trying to see if it's possible to change the Popover/Portal code to not create div elements until the Popover is actually opened.
This was another suggestion I was thinking of last night, but I'm not 100% sure if this will break things (especially worried about LegacyPopover, but ideally we would finish migrating those, so maybe if it does break those we could just do that first).

If we get to a place where this stuff only renders when it's "in use" then I'm cool with that. I think it might be pretty hard to do that without the previous mentioned approaches though, purely because of the close animations (which effectively require the transition parent to stay mounted). Perhaps react-spring allows us to remove the parent post-animation easily, though?